File: shared/src/commonTest/kotlin/space/denhan/domain/RecurrenceTest.kt
package space.denhan.domain

import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.LocalDate
import kotlinx.datetime.plus
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class RecurrenceTest {

    private fun d(iso: String) = LocalDate.parse(iso)

    // The bug this whole file exists for. Adding a month at a time from 31 Jan
    // permanently loses the 31st; anchoring does not.
    @Test
    fun `monthly from the 31st returns to the 31st instead of drifting`() {
        val anchor = d("2026-01-31")

        assertEquals(d("2026-02-28"), Recurrence.occurrenceAfter(anchor, Cycle.MONTHLY, 1))
        assertEquals(d("2026-03-31"), Recurrence.occurrenceAfter(anchor, Cycle.MONTHLY, 2))
        assertEquals(d("2026-04-30"), Recurrence.occurrenceAfter(anchor, Cycle.MONTHLY, 3))
        assertEquals(d("2026-05-31"), Recurrence.occurrenceAfter(anchor, Cycle.MONTHLY, 4))
    }

    @Test
    fun `accumulating one month at a time is what we must not do`() {
        // Documents the wrong approach so the difference stays visible.
        var drifting = d("2026-01-31")
        repeat(4) { drifting = drifting.plus(1, DateTimeUnit.MONTH) }

        assertEquals(d("2026-05-28"), drifting, "accumulation loses the 31st")
        assertEquals(
            d("2026-05-31"),
            Recurrence.occurrenceAfter(d("2026-01-31"), Cycle.MONTHLY, 4),
            "anchoring keeps it",
        )
    }

    @Test
    fun `29 February resolves in a leap year and clamps otherwise`() {
        assertEquals(d("2028-02-29"), Recurrence.occurrenceAfter(d("2024-02-29"), Cycle.YEARLY, 4))
        assertEquals(d("2025-02-28"), Recurrence.occurrenceAfter(d("2024-02-29"), Cycle.YEARLY, 1))
    }

    @Test
    fun `occurrence zero is the anchor itself`() {
        val anchor = d("2026-08-15")
        assertEquals(anchor, Recurrence.occurrenceAfter(anchor, Cycle.MONTHLY, 0))
    }

    @Test
    fun `weekly steps seven days`() {
        assertEquals(d("2026-08-29"), Recurrence.occurrenceAfter(d("2026-08-15"), Cycle.WEEKLY, 2))
    }

    @Test
    fun `quarterly and yearly step the right number of months`() {
        val anchor = d("2026-01-15")
        assertEquals(d("2026-04-15"), Recurrence.occurrenceAfter(anchor, Cycle.QUARTERLY, 1))
        assertEquals(d("2026-07-15"), Recurrence.occurrenceAfter(anchor, Cycle.SEMIANNUAL, 1))
        assertEquals(d("2027-01-15"), Recurrence.occurrenceAfter(anchor, Cycle.YEARLY, 1))
    }

    @Test
    fun `cyclesElapsed counts whole cycles only`() {
        val anchor = d("2026-01-15")
        assertEquals(0, Recurrence.cyclesElapsed(anchor, Cycle.MONTHLY, d("2026-01-15")))
        assertEquals(0, Recurrence.cyclesElapsed(anchor, Cycle.MONTHLY, d("2026-02-14")))
        assertEquals(1, Recurrence.cyclesElapsed(anchor, Cycle.MONTHLY, d("2026-02-15")))
        assertEquals(7, Recurrence.cyclesElapsed(anchor, Cycle.MONTHLY, d("2026-08-15")))
    }

    @Test
    fun `cyclesElapsed is zero before the anchor`() {
        assertEquals(0, Recurrence.cyclesElapsed(d("2026-08-15"), Cycle.MONTHLY, d("2025-01-01")))
    }

    @Test
    fun `nextDue returns the anchor while it is still ahead`() {
        assertEquals(
            d("2026-09-01"),
            Recurrence.nextDue(d("2026-09-01"), Cycle.MONTHLY, d("2026-08-15")),
        )
    }

    @Test
    fun `nextDue steps past today for a running cycle`() {
        assertEquals(
            d("2026-09-15"),
            Recurrence.nextDue(d("2026-01-15"), Cycle.MONTHLY, d("2026-08-15")),
        )
    }

    @Test
    fun `nextDue on the due date itself moves to the following cycle`() {
        assertEquals(
            d("2026-09-15"),
            Recurrence.nextDue(d("2026-01-15"), Cycle.MONTHLY, d("2026-08-15")),
        )
    }

    @Test
    fun `nextDue for a one-off is null once it has passed`() {
        assertNull(Recurrence.nextDue(d("2026-01-01"), null, d("2026-08-15")))
        assertEquals(d("2026-12-01"), Recurrence.nextDue(d("2026-12-01"), null, d("2026-08-15")))
    }

    @Test
    fun `actBy subtracts the lead so the deadline is actionable`() {
        assertEquals(d("2026-08-08"), Recurrence.actBy(d("2026-08-15"), 7))
        assertEquals(d("2026-08-15"), Recurrence.actBy(d("2026-08-15"), 0))
    }

    @Test
    fun `actBy crosses a month boundary correctly`() {
        assertEquals(d("2026-07-31"), Recurrence.actBy(d("2026-08-01"), 1))
    }
}

File: test/unit/domain/recurrence_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:subdock/domain/local_date.dart';
import 'package:subdock/domain/recurrence.dart';

void main() {
  LocalDate d(String iso) => LocalDate.parse(iso);

  // The bug this whole file exists for. Adding a month at a time from 31 Jan
  // permanently loses the 31st; anchoring does not.
  test('monthly from the 31st returns to the 31st instead of drifting', () {
    final anchor = d('2026-01-31');

    expect(
      Recurrence.occurrenceAfter(anchor, Cycle.monthly, 1),
      d('2026-02-28'),
    );
    expect(
      Recurrence.occurrenceAfter(anchor, Cycle.monthly, 2),
      d('2026-03-31'),
    );
    expect(
      Recurrence.occurrenceAfter(anchor, Cycle.monthly, 3),
      d('2026-04-30'),
    );
    expect(
      Recurrence.occurrenceAfter(anchor, Cycle.monthly, 4),
      d('2026-05-31'),
    );
  });

  test('accumulating one month at a time is what we must not do', () {
    // Documents the wrong approach so the difference stays visible.
    var drifting = d('2026-01-31');
    for (var i = 0; i < 4; i++) {
      drifting = drifting.plusMonths(1);
    }

    expect(drifting, d('2026-05-28'), reason: 'accumulation loses the 31st');
    expect(
      Recurrence.occurrenceAfter(d('2026-01-31'), Cycle.monthly, 4),
      d('2026-05-31'),
      reason: 'anchoring keeps it',
    );
  });

  test('29 February resolves in a leap year and clamps otherwise', () {
    expect(
      Recurrence.occurrenceAfter(d('2024-02-29'), Cycle.yearly, 4),
      d('2028-02-29'),
    );
    expect(
      Recurrence.occurrenceAfter(d('2024-02-29'), Cycle.yearly, 1),
      d('2025-02-28'),
    );
  });

  test('occurrence zero is the anchor itself', () {
    final anchor = d('2026-08-15');
    expect(Recurrence.occurrenceAfter(anchor, Cycle.monthly, 0), anchor);
  });

  test('a negative occurrence is rejected rather than walking backwards', () {
    expect(
      () => Recurrence.occurrenceAfter(d('2026-08-15'), Cycle.monthly, -1),
      throwsA(isA<ArgumentError>()),
    );
  });

  test('weekly steps seven days', () {
    expect(
      Recurrence.occurrenceAfter(d('2026-08-15'), Cycle.weekly, 2),
      d('2026-08-29'),
    );
  });

  test('quarterly and yearly step the right number of months', () {
    final anchor = d('2026-01-15');
    expect(
      Recurrence.occurrenceAfter(anchor, Cycle.quarterly, 1),
      d('2026-04-15'),
    );
    expect(
      Recurrence.occurrenceAfter(anchor, Cycle.semiannual, 1),
      d('2026-07-15'),
    );
    expect(
      Recurrence.occurrenceAfter(anchor, Cycle.yearly, 1),
      d('2027-01-15'),
    );
  });

  test('cyclesElapsed counts whole cycles only', () {
    final anchor = d('2026-01-15');
    expect(Recurrence.cyclesElapsed(anchor, Cycle.monthly, d('2026-01-15')), 0);
    expect(Recurrence.cyclesElapsed(anchor, Cycle.monthly, d('2026-02-14')), 0);
    expect(Recurrence.cyclesElapsed(anchor, Cycle.monthly, d('2026-02-15')), 1);
    expect(Recurrence.cyclesElapsed(anchor, Cycle.monthly, d('2026-08-15')), 7);
  });

  test('cyclesElapsed is zero before the anchor', () {
    expect(
      Recurrence.cyclesElapsed(d('2026-08-15'), Cycle.monthly, d('2025-01-01')),
      0,
    );
  });

  test('nextDue returns the anchor while it is still ahead', () {
    expect(
      Recurrence.nextDue(d('2026-09-01'), Cycle.monthly, d('2026-08-15')),
      d('2026-09-01'),
    );
  });

  test('nextDue steps past today for a running cycle', () {
    expect(
      Recurrence.nextDue(d('2026-01-15'), Cycle.monthly, d('2026-08-15')),
      d('2026-09-15'),
    );
  });

  test('nextDue on the due date itself moves to the following cycle', () {
    expect(
      Recurrence.nextDue(d('2026-08-15'), Cycle.monthly, d('2026-08-15')),
      d('2026-09-15'),
    );
  });

  test('nextDue for a one-off is null once it has passed', () {
    expect(Recurrence.nextDue(d('2026-01-01'), null, d('2026-08-15')), isNull);
    expect(
      Recurrence.nextDue(d('2026-12-01'), null, d('2026-08-15')),
      d('2026-12-01'),
    );
  });

  test('actBy subtracts the lead so the deadline is actionable', () {
    expect(Recurrence.actBy(d('2026-08-15'), 7), d('2026-08-08'));
    expect(Recurrence.actBy(d('2026-08-15'), 0), d('2026-08-15'));
  });

  test('actBy crosses a month boundary correctly', () {
    expect(Recurrence.actBy(d('2026-08-01'), 1), d('2026-07-31'));
  });

  test('a negative act-by offset is rejected', () {
    expect(
      () => Recurrence.actBy(d('2026-08-15'), -1),
      throwsA(isA<ArgumentError>()),
    );
  });
  test('occurrenceBefore steps back the same number of cycles', () {
    expect(
      Recurrence.occurrenceBefore(d('2026-08-20'), Cycle.monthly, 3),
      d('2026-05-20'),
    );
    expect(
      Recurrence.occurrenceBefore(d('2026-08-20'), Cycle.weekly, 2),
      d('2026-08-06'),
    );
    expect(
      Recurrence.occurrenceBefore(d('2026-08-20'), Cycle.yearly, 1),
      d('2025-08-20'),
    );
  });

  // Stepping back from a date that was itself clamped does not land on the
  // day it was clamped from. The date the user typed is the one worth keeping,
  // so the loss is documented rather than worked around.
  test('occurrenceBefore is not an exact inverse across a clamped month', () {
    final clamped = Recurrence.occurrenceAfter(
      d('2026-01-31'),
      Cycle.monthly,
      1,
    );
    expect(clamped, d('2026-02-28'));
    expect(
      Recurrence.occurrenceBefore(clamped, Cycle.monthly, 1),
      d('2026-01-28'),
    );
  });

  test('a negative step back is rejected', () {
    expect(
      () => Recurrence.occurrenceBefore(d('2026-08-15'), Cycle.monthly, -1),
      throwsA(isA<ArgumentError>()),
    );
  });
}
